src/analytics/analytics.js

```diff
@@ -0,0 +1,155 @@
+/** @flow */
+import serializeError from 'serialize-error';
+import requestify from 'requestify';
+import hashObj from 'object-hash';
+import uniqid from 'uniqid';
+import yn from 'yn';
+import R from 'ramda';
+import omitBy from 'lodash.omitby';
+import isNil from 'lodash.isnil';
+import logger from '../logger/logger';
+import { setSync, getSync } from '../api/consumer/lib/global-config';
+import {
+  CFG_ANALYTICS_USERID_KEY,
+  CFG_ANALYTICS_REPORTING_KEY,
+  CFG_ANALYTICS_ERROR_REPORTS_KEY,
+  CFG_ANALYTICS_ANONYMOUS_KEY,
+  CFG_USER_EMAIL_KEY,
+  CFG_USER_NAME_KEY,
+  DEFAULT_BIT_ENV,
+  CFG_ANALYTICS_ENVIRONMENT_KEY,
+  ANALYTICS_URL
+} from '../constants';
+
+const LEVEL = {
+  DEBUG: 'debug',
+  INFO: 'info',
+  WARNING: 'warning',
+  ERROR: 'error',
+  FATAL: 'fatal'
+};
+
+class Breadcrumb {
+  message: string;
+  data: Object;
+
+  constructor(category: string, message: string, data: Object) {
+    this.category = category;
+    this.message = message;
+    this.data = data;
+  }
+}
+class Analytics {
+  static username: string;
+  static command: string;
+  static release: string;
+  static args: string[];
+  static flags: Object = {};
+  static success: boolean = true;
+  static nodeVersion: string;
+  static os: string;
+  static extra: ?Object = {};
+  static level: string;
+  static error: Error | string | Object;
+  static breadcrumbs: Array<Breadcrumb> = [];
+  static analytics_usage: boolean;
+  static error_usage: boolean;
+  static anonymous: boolean;
+  static environment: string;
+
+  static getID(): string {
+    const id = getSync(CFG_ANALYTICS_USERID_KEY);
+    const newId = uniqid();
+    if (id) return id;
+    setSync(CFG_ANALYTICS_USERID_KEY, newId);
+    return newId;
+  }
+
+  static _hashArgs(flags: Object) {
+    const hashedFlags = {};
+    const filteredFlags = omitBy(flags, isNil);
+    if (this.anonymous && !R.isEmpty(filteredFlags)) {
+      Object.keys(filteredFlags).forEach((key) => {
+        if (typeof flags[key] === 'boolean') {
+          hashedFlags[key] = flags[key];
+        } else {
+          hashedFlags[key] = hashObj(flags[key]);
+        }
+      });
+      return hashedFlags;
+    }
+    return filteredFlags;
+  }
+  static init(command: string, flags: Object, args: string[], version) {
+    this.anonymous = yn(getSync(CFG_ANALYTICS_ANONYMOUS_KEY), { default: true });
+    this.command = command;
+    this.flags = this._hashArgs(flags);
+    this.release = version;
+    this.args = this.anonymous ? hashObj(args) : args;
+    this.nodeVersion = process.version;
+    this.os = process.platform;
+    this.level = LEVEL.INFO;
+    this.username = !this.anonymous
+      ? getSync(CFG_USER_NAME_KEY) || getSync(CFG_USER_EMAIL_KEY) || this.getID()
```

Comment:
I think it's better to take the email before username. also maybe before the getID we can add the machine name?